Make gossip reactive instead of only at fixed intervals.

spencerkimball opened the pull request:
New infos are now gossiped as soon as they're set instead of
at fixed intervals. Removed the info sequence number in favor
of an info timestamp, set at the originating node. Requests
for new gossip from a server are now accompanied by the
requesting node's set of high water timestamps (that is, a map
of node IDs to the most recently info timestamp seen for each
node). All gossip responses contain the responding node's
latest high water timestamp map.

Gossip clients now send two requests to servers: one is a
"hanging" get which waits for the server to have new info.
The other is an info "push", which is triggered any time the
node receives new gossip of its own.

Added new callback mechanism to gossip which allows interested
callers to specify a channel to receive gossip updates instead
of a callback method.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/2967)

<!-- Reviewable:end -->

spencerkimball committed:
Make gossip reactive instead of only at fixed intervals.

New infos are now gossiped as soon as they're set instead of
at fixed intervals. Removed the info sequence number in favor
of an info timestamp, set at the originating node. Requests
for new gossip from a server are now accompanied by the
requesting node's set of high water timestamps (that is, a map
of node IDs to the most recently info timestamp seen for each
node). All gossip responses contain the responding node's
latest high water timestamp map.

Gossip clients now send two requests to servers: one is a
"hanging" get which waits for the server to have new info.
The other is an info "push", which is triggered any time the
node receives new gossip of its own.

Added callback unregister mechanism.

Fixed and tested gossip network simulation.

tschottdorf commented on gossip/client.go:
```diff
@@ -147,28 +156,59 @@ func (c *client) gossip(g *Gossip, stopper *stop.Stopper) error {
+	// server when this node has new gossip information to share with
+	// the server.
+	for {
+		select {
```
what's the connection timeout handling here?

spencerkimball replied:
You mean where did it go? I'm figuring we just skip those here. Other parts of the system are already monitoring our links for failures and timeouts and will kill them. If the rpc is closed out from under the gossip client, it does the right thing. The old timeout code was redundant.

tschottdorf commented on gossip/client_test.go:
```diff
@@ -111,49 +105,3 @@ func TestClientGossip(t *testing.T) {
-// TestClientDisconnectRedundant verifies that the gossip server
-// will drop an outgoing client connection that is already an
-// inbound client connection of another node.
-func TestClientDisconnectRedundant(t *testing.T) {
```
is this test really obsolete now?

spencerkimball replied:
No, I added the code to delete a cycle back in and restored this test.

tamird commented:
This assumes that all node clocks are monotonic, right? Should we be using HLC timestamps to preserve this invariant?
<!-- Reviewable comment -K1u-oMxtOOng-szGPkj:2068417626 -->

---

Reviewed 24 of 26 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 38 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tWLZe3GWVQXOfOzXY)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L38)):</sup>
Can you add a comment explaining why this key type can't be `roachpb.NodeID`? Either here or in the proto.
<!-- Reviewable comment -K1tWLZe3GWVQXOfOzXZ:-1063325672 -->

---

<sup>**[gossip/client.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u16MceGg3RfJUSZ4D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L46)):</sup>
this buffer size worries me because it's arbitrary. what if you made this channel unbuffered?
<!-- Reviewable comment -K1u16MceGg3RfJUSZ4E:1815198667 -->

---

<sup>**[gossip/client.go, line 58 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1thSXygxTfDR_EBiGB)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L58)):</sup>
I'd use the exported version of this method here.
<!-- Reviewable comment -K1thSXygxTfDR_EBiGC:-1609914170 -->

---

<sup>**[gossip/client.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1Dcb6MkGd6iJT7-A)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L76)):</sup>
I'd use the exported version of this method here.
<!-- Reviewable comment -K1u1Dcb6MkGd6iJT7-B:-1609914170 -->

---

<sup>**[gossip/client.go, line 95 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1ITELG3jcfXHSq4X)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L95)):</sup>
The way this method is being used suggests that it should take a channel argument and not return anything.
<!-- Reviewable comment -K1u1ITELG3jcfXHSq4Y:-960089490 -->

---

<sup>**[gossip/client.go, line 111 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1MxNDaZuCS7JPKu7)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L111)):</sup>
The way this method is being used suggests that it should take a channel argument and not return anything.
<!-- Reviewable comment -K1u1MxNDaZuCS7JPKu8:-960089490 -->

---

<sup>**[gossip/client.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1WLDXzQZgBQLFQY6)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L133)):</sup>
This locking could happen a bit lower down; no need to hold the lock during some of this error checking.
<!-- Reviewable comment -K1u1WLDXzQZgBQLFQY7:-324313528 -->

---

<sup>**[gossip/client.go, line 193 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tVUErWp_PTru0r91P-r2-193)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L193)):</sup>
It's all async; what timeout handling would you expect?
<!-- Reviewable comment -K1u59YbIOShkauJpl15:-101290852 -->

---

<sup>**[gossip/client.go, line 199 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u0ka7lYBtJ44sOtyS)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L199)):</sup>
Instead of listening on these channels, you should create a single channel that you pass to getGossip and sendGossip
<!-- Reviewable comment -K1u0ka7lYBtJ44sOtyT:169968964 -->

---

<sup>**[gossip/gossip.go, line 31 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u2M6lzY0Mrxh_ifz3)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.go#L31)):</sup>
s/info/Info/
<!-- Reviewable comment -K1u2M6lzY0Mrxh_ifz4:-1033862879 -->

---

<sup>**[gossip/gossip.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u43hXTk7rUhaIcNGG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.go#L325)):</sup>
File an issue when this is merged.
<!-- Reviewable comment -K1u43hXTk7rUhaIcNGH:485701731 -->

---

<sup>**[gossip/gossip.proto, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1typfJfQYAmQEyjtNB)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.proto#L44)):</sup>
I don't think you need this custom name.
<!-- Reviewable comment -K1typfKrMiGYV-t7maN:1140229127 -->

---

<sup>**[gossip/gossip.proto, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tysCvYNQogF-0KcYX)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.proto#L62)):</sup>
or this one.
<!-- Reviewable comment -K1tysCvYNQogF-0KcYY:1883462226 -->

---

<sup>**[gossip/gossip.proto, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tz86sYRDJ4tv1-Hen)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.proto#L69)):</sup>
Something off about this sentence.
<!-- Reviewable comment -K1tz86sYRDJ4tv1-Heo:594675448 -->

---

<sup>**[gossip/info.go, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tzWKN0HQF1OctJOy4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8b54f183cea3404270b03b1d36c4847a10354829/gossip/info.go#L45)):</sup>
what's the implication of removing this?
<!-- Reviewable comment -K1tzWKOSqAmKQTeV2wD:-652830548 -->

---

<sup>**[gossip/infostore.go, line 213 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u4kVOe_yxO82qOw1G)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore.go#L213)):</sup>
We should consider sending `i.Value` on the channel
<!-- Reviewable comment -K1u4bvUtpYoVFrAj9ry:1064737196 -->

---

<sup>**[gossip/infostore.go, line 228 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u4xNZUFry_TCWDBEl)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore.go#L228)):</sup>
the `0` is redundant here
<!-- Reviewable comment -K1u4xNZUFry_TCWDBEm:924979093 -->

---

<sup>**[gossip/infostore_test.go, line 226 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tsRHY_O_B0XRy4kjR)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore_test.go#L226)):</sup>
"nil" isn't quite right, and we should log the entire `infos` here.
<!-- Reviewable comment -K1tsRHY_O_B0XRy4kjS:-1206970434 -->

---

<sup>**[gossip/infostore_test.go, line 340 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1ttPVWNgGrPfAoFg53)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore_test.go#L340)):</sup>
can you `range` over a slice of 3 structs here to avoid repeating yourself?
<!-- Reviewable comment -K1ttPVWNgGrPfAoFg54:-944882571 -->

---

<sup>**[gossip/infostore_test.go, line 385 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tty3dfd6DdH8o1jHZ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore_test.go#L385)):</sup>
is there order here not deterministic? wondering why you need to (and why it's ok to) sort.
<!-- Reviewable comment -K1tty3eepcGQ_OyWH87:390378812 -->

---

<sup>**[gossip/server.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3l_IU0qULCSPAOWt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L56)):</sup>
same comment about the size of this buffer
<!-- Reviewable comment -K1u3l_IU0qULCSPAOWu:-965808128 -->

---

<sup>**[gossip/server.go, line 121 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3f9oClXJXXHHLLux)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L121)):</sup>
Don't you also want to do this opportunistically above?
<!-- Reviewable comment -K1u3f9oClXJXXHHLLuy:1722177766 -->

---

<sup>**[kv/dist_sender_test.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tyazPIYIG6O9zeavK)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/kv/dist_sender_test.go#L56)):</sup>
This comment is not totally accurate; this interval is the "adding info interval" now.
<!-- Reviewable comment -K1tyazQTKCoeW2vtt92:903062065 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 193 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tVUErWp_PTru0r91P-r2-193)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L193)):</sup>
haven't looked into it closely, just want to point out that in reality the RPC could time out (?) before there's activity on the line, which would not constitute an error.
<!-- Reviewable comment -K1u6OXICKUqfP1zpeyb:1264643907 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 193 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tVUErWp_PTru0r91P-r2-193)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L193)):</sup>
Yeah; let's circle back to this once the channel-passing stuff is refactored here.
<!-- Reviewable comment -K1u7jJSwxUo3875TjC5:-463014146 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client_test.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u8ss_luYgr55no86X)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8b54f183cea3404270b03b1d36c4847a10354829/gossip/client_test.go#L118)):</sup>
+1 to tobi's comment; this needs to come back.
<!-- Reviewable comment -K1u8ssaemFrdkIV5cAd:-457726579 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 26 files at r1.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client_test.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u8ss_luYgr55no86X)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8b54f183cea3404270b03b1d36c4847a10354829/gossip/client_test.go#L118)):</sup>
Nope, disregard. This functionality was removed. The method `(*Gossip).hasIncoming` is now unused and should be removed, though.
<!-- Reviewable comment -K1uDy66-tOFd_HNfZyy:1117151976 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 194 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1uJ-fTF5cIKQW-2cLn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L194)):</sup>
I do not think that connections need to be push-pull; they should be one or the other, probably just 'push'.  A 'pull' request just does a long poll, and doesn't return until there is new information on the other side of the connection; therefore, a push from the other direction would be unnecessary.  The same thing goes for a 'push' request: since a push always occurs as soon as new information is available, there's no reason for the other side of the connection to be long-polling in the first place.
<!-- Reviewable comment -K1uJ-fTF5cIKQW-2cLo:-117833362 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

spencerkimball commented:
<sup>**[gossip/client.go, line 38 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tWLZe3GWVQXOfOzXY)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L38)):</sup>
Added.
<!-- Reviewable comment -K1vV27J4gwr31bqalEI:1610000531 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 24 of 27 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u16MceGg3RfJUSZ4D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L46)):</sup>
then adding an info to an infostore would potentially block up a lot of other stuff.
<!-- Reviewable comment -K1vVTG9AcAQBh3SAweO:-757351506 -->

---

<sup>**[gossip/client.go, line 58 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1thSXygxTfDR_EBiGB)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L58)):</sup>
The caller has the mutex held here. Added a comment.
<!-- Reviewable comment -K1vVcG9vHLWOWITNLfe:-1448419644 -->

---

<sup>**[gossip/client.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1Dcb6MkGd6iJT7-A)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L76)):</sup>
We could do that but since we're using the unexported version above, better to be consistent here.
<!-- Reviewable comment -K1vVfEIig9U9xZKy2qE:-1069812849 -->

---

<sup>**[gossip/client.go, line 95 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1ITELG3jcfXHSq4X)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L95)):</sup>
Changed.
<!-- Reviewable comment -K1vW1HD9FoTHA5JpSPJ:-609445857 -->

---

<sup>**[gossip/client.go, line 111 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1MxNDaZuCS7JPKu7)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L111)):</sup>
Changed.
<!-- Reviewable comment -K1vYU-LdV56LXNmnDSt:-609445857 -->

---

<sup>**[gossip/client.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1WLDXzQZgBQLFQY6)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L133)):</sup>
Errors are the exception case, not the common case. I prefer to put the locking at the start of methods if possible as it is less error prone as the code changes.
<!-- Reviewable comment -K1vW5b_lZR-nO5_Pb09:-320376186 -->

---

<sup>**[gossip/client.go, line 194 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1uJ-fTF5cIKQW-2cLn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L194)):</sup>
I tried doing it just one direction, but then there's no impetus for network forming. You need to pull in order to get to the sentinel gossip. If you don't also push, then there's no guarantee anyone else will ever get your gossip.
<!-- Reviewable comment -K1vWN1M7siThu6HVpqF:267697948 -->

---

<sup>**[gossip/client.go, line 199 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u0ka7lYBtJ44sOtyS)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L199)):</sup>
Gotcha. Good idea. Changed.
<!-- Reviewable comment -K1vYRHCio2pfxMWDTlc:-589445665 -->

---

<sup>**[gossip/gossip.go, line 31 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u2M6lzY0Mrxh_ifz3)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.go#L31)):</sup>
OK
<!-- Reviewable comment -K1vYei8Rs-sKjxdRHKS:5173665 -->

---

<sup>**[gossip/gossip.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u43hXTk7rUhaIcNGG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.go#L325)):</sup>
OK
<!-- Reviewable comment -K1vYjQiKl0t7LNJF37f:5173665 -->

---

<sup>**[gossip/gossip.proto, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1typfJfQYAmQEyjtNB)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.proto#L44)):</sup>
Good point.
<!-- Reviewable comment -K1vYlQqM7eAYNQerL3h:807670556 -->

---

<sup>**[gossip/gossip.proto, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tysCvYNQogF-0KcYX)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.proto#L62)):</sup>
Changed.
<!-- Reviewable comment -K1vYsrCX66bD2OQn7s2:-609445857 -->

---

<sup>**[gossip/gossip.proto, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tz86sYRDJ4tv1-Hen)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.proto#L69)):</sup>
Fixed.
<!-- Reviewable comment -K1vYxTjupa7J2tx68qI:1758359711 -->

---

<sup>**[gossip/infostore.go, line 213 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u4kVOe_yxO82qOw1G)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore.go#L213)):</sup>
It's not returned elsewhere, notably when getInfo is called.
<!-- Reviewable comment -K1vZ0H12l66ftMkZrvC:-1551001380 -->

---

<sup>**[gossip/infostore.go, line 228 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u4xNZUFry_TCWDBEl)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore.go#L228)):</sup>
Removed.
<!-- Reviewable comment -K1vZ6e89xgYtHQ9pTwc:1740281843 -->

---

<sup>**[gossip/infostore_test.go, line 226 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tsRHY_O_B0XRy4kjR)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore_test.go#L226)):</sup>
Changed to empty.
<!-- Reviewable comment -K1vZEEweDkYu37dj13a:-489080619 -->

---

<sup>**[gossip/infostore_test.go, line 340 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1ttPVWNgGrPfAoFg53)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore_test.go#L340)):</sup>
it's a test...
<!-- Reviewable comment -K1vZGjB6P-DKu2P6zRr:-442784039 -->

---

<sup>**[gossip/infostore_test.go, line 385 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tty3dfd6DdH8o1jHZ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/infostore_test.go#L385)):</sup>
anachronism from copy and paste. Removed.
<!-- Reviewable comment -K1vZLDy3mZZH0454_om:1246222203 -->

---

<sup>**[gossip/server.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3l_IU0qULCSPAOWt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L56)):</sup>
I don't see a better option.
<!-- Reviewable comment -K1vZaDAS2nSMX71oLhl:-2146294751 -->

---

<sup>**[gossip/server.go, line 121 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3f9oClXJXXHHLLux)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L121)):</sup>
No need. The channel will already have any pre-existing infos waiting on it.
<!-- Reviewable comment -K1vZeJT4oPn0nienM0N:618832855 -->

---

<sup>**[kv/dist_sender_test.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tyazPIYIG6O9zeavK)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/kv/dist_sender_test.go#L56)):</sup>
Removed.
<!-- Reviewable comment -K1vZlwgCOg8GUx_QZL1:1740281843 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r3, 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u1Dcb6MkGd6iJT7-A)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L76)):</sup>
I don't agree, but OK.
<!-- Reviewable comment -K1v_92dJaPDqknpnDk1:-332482113 -->

---

<sup>**[gossip/client.go, line 194 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1uJ-fTF5cIKQW-2cLn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L194)):</sup>
Why do you need to pull in order to get to the sentinel gossip? If you allow bidirectional connections, someone will eventually connect back to you after your push makes them aware of your existence.
<!-- Reviewable comment -K1v_K5zQ1k6psrkwPBk:-1228697764 -->

---

<sup>**[gossip/client.go, line 193 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1vbBH44ktAVTHF_xaN)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/client.go#L193)):</sup>
this doesn't need to be this big, does it? from what i understand it's expected that we usually have at most 2 outstanding RPCs; this should be more like 10 rather than 10000.
<!-- Reviewable comment -K1vbBH5mRHi_pMJI90p:1414080024 -->

---

<sup>**[gossip/gossip.go, line 458 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1vaf5QBOr-z_Qf0jS-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.go#L458)):</sup>
how come this doesn't need to be under lock?
<!-- Reviewable comment -K1vaf5QBOr-z_Qf0jS0:1924096739 -->

---

<sup>**[gossip/infostore_test.go, line 226 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1va2NphAM49xUAWF7-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/infostore_test.go#L226)):</sup>
can you log `infos` instead of `len(infos)`?
<!-- Reviewable comment -K1va2NphAM49xUAWF70:960956418 -->

---

<sup>**[gossip/infostore_test.go, line 417 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1va8FMBu-FH7sQ-5nZ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/infostore_test.go#L417)):</sup>
my questions stands: why is it ok to sort these?
<!-- Reviewable comment -K1va8FMBu-FH7sQ-5n_:-1606773311 -->

---

<sup>**[gossip/server.go, line 121 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3f9oClXJXXHHLLux)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L121)):</sup>
I don't follow; which channel?
<!-- Reviewable comment -K1vgPFkkjVZx2MdMdKd:1729341857 -->

---

<sup>**[gossip/simulation/gossip.go, line 201 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1vaST30_MVwJ8vJFsV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/simulation/gossip.go#L201)):</sup>
it may be easier to change this to a uint64 rather than sprinkle all those conversions below
<!-- Reviewable comment -K1vaST30_MVwJ8vJFsW:507208609 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u16MceGg3RfJUSZ4D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L46)):</sup>
Why would it? This channel is just the callback channel; adding an info to an infostore signals it, which immediately results in an outgoing rpc (async), which is fast. What's blocking?
<!-- Reviewable comment -K1vjDR0x50TPxhYgSNx:-548037166 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/server.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3l_IU0qULCSPAOWt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L56)):</sup>
I don't see why this needs to be buffered; this channel has a dedicated goroutine servicing it that does nothing more than call `Broadcast()` on a condvar each time there's an update.
<!-- Reviewable comment -K1vjdkUhq8W8MO2JIIi:512422496 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u16MceGg3RfJUSZ4D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L46)):</sup>
There are some tricky locking issues with handling the responses from push and pulls when combining the deltas, which in turn send updates on channel. If the channel blocks, things get wedged. I can avoid the issue but the complexity of the code will increase non trivially.
<!-- Reviewable comment -K29WFm3TzJ_Tvxg9gDX:-1565566806 -->

---

<sup>**[gossip/client.go, line 194 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1uJ-fTF5cIKQW-2cLn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L194)):</sup>
Not sure I understand your comment. We are allowing bidirectional connections. We push and pull from the client. We pull in order to join the network and then to form connections to "distant" nodes.
<!-- Reviewable comment -K2855zmYR-1VyU5GVD_:1519515338 -->

---

<sup>**[gossip/client.go, line 193 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1vbBH44ktAVTHF_xaN)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/client.go#L193)):</sup>
We can have as many RPCs outstanding on the push side as infos have been added to infostore. Could be quite a number.
<!-- Reviewable comment -K284mSbVZQnCio5hdec:1149192730 -->

---

<sup>**[gossip/gossip.go, line 458 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1vaf5QBOr-z_Qf0jS-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/gossip.go#L458)):</sup>
It does. Lock is already held when called. Added comment.
<!-- Reviewable comment -K285malw3R7tD_N19s6:-1093134211 -->

---

<sup>**[gossip/info.go, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1tzWKN0HQF1OctJOy4)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8b54f183cea3404270b03b1d36c4847a10354829/gossip/info.go#L45)):</sup>
It was a small optimization before. It's no longer useful with the high water timestamps.
<!-- Reviewable comment -K286AFwy3JT_zoUfych:-561210683 -->

---

<sup>**[gossip/infostore_test.go, line 226 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1va2NphAM49xUAWF7-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/infostore_test.go#L226)):</sup>
OK
<!-- Reviewable comment -K286T2vEjFSpm75zNaM:5173665 -->

---

<sup>**[gossip/infostore_test.go, line 417 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1va8FMBu-FH7sQ-5nZ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/infostore_test.go#L417)):</sup>
Because they're sent on the channel from a map which has an undefined ordering on iteration.
<!-- Reviewable comment -K286bHDnNaWeGl5oI1y:844873465 -->

---

<sup>**[gossip/server.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3l_IU0qULCSPAOWt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L56)):</sup>
In this case I can remove the buffering and launch the initial channel registration in a goroutine.
<!-- Reviewable comment -K29Yxlb4kxiAW5VE_e5:1130539987 -->

---

<sup>**[gossip/server.go, line 121 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3f9oClXJXXHHLLux)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L121)):</sup>
The client handling code expects one or the other sort of request. I don't want to confuse things unnecessarily. Either it's a push which doesn't receive a delta or a pull which does.
<!-- Reviewable comment -K29_-JkS3de896bjyGJ:620444367 -->

---

<sup>**[gossip/simulation/gossip.go, line 201 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1vaST30_MVwJ8vJFsV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/8657c36f0d35ddd771a5123c8d1bdce3ab198c9b/gossip/simulation/gossip.go#L201)):</sup>
Get complaints about subtraction if we do.
<!-- Reviewable comment -K29_JADSjc6fUKplt9g:1601947660 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u16MceGg3RfJUSZ4D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L46)):</sup>
This makes me think this isn't the right pattern; we've seen issues with arbitrarily buffered channels before. Maybe callbacks are the right thing after all? cc @bdarnell
<!-- Reviewable comment -K2AAR3YYJptmq7OeOXO:596366698 -->

---

<sup>**[gossip/server.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u3l_IU0qULCSPAOWt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/server.go#L56)):</sup>
Did you mean to make that change?
<!-- Reviewable comment -K2AF4BdRVrNQY0i5V34:-930389001 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K1u16MceGg3RfJUSZ4D)** ([raw file](https://github.com/cockroachdb/cockroach/blob/6df37958cb6b52df4376aafffa8e534a7d0b43fd/gossip/client.go#L46)):</sup>
Yeah, it's not a good idea to give a channel a large buffer to avoid deadlocks. If deadlocks are a possibility they will generally be possible for any finite buffer size. If callbacks work I think that would be preferable here, but callbacks can have similar issues to unbuffered channels, since a callback can also block the thread. 

If there's no safe place for a consumer to apply backpressure then you may have to do what we did in multiraft (#2448) and send slices over the channel instead of individual elements (simulating a channel that is bounded only by available memory).
<!-- Reviewable comment -K2AHMC7VHRR55a6M3pJ:2074266988 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@tamird PTAL. I removed the large buffered channels and went back to using the callback interface.

tamird commented:
LGTM. Let's squash this before merging. One nit about the UnregisterCallback interface: we've started adopting the practice of returning a so-called undoer callback in other places - here that would be this diff that I tested locally: https://gist.github.com/tamird/6a2061cd7bb48dde1187
<!-- Reviewable comment -K2EWReTuIIirXG5WLsB:-1594883803 -->

---

Reviewed 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Changed.
<!-- Reviewable comment -K2HvQONT8nUX4nHra2r:-609445857 -->

---

Review status: 23 of 27 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 5 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/client.go, line 190 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2IPfytDTmZnqEJ5ufa)** ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fdc51c55e9e18c7efec4cbf22b4673a110a2fa/gossip/client.go#L190)):</sup>
s/calblack/callback/
<!-- Reviewable comment -K2IPfyu_qbas8p50LVv:959403149 -->

---

<sup>**[gossip/server.go, line 169 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2IPmCDPFzBM2gV5Z5h)** ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fdc51c55e9e18c7efec4cbf22b4673a110a2fa/gossip/server.go#L169)):</sup>
nit (which I'm only leaving because there's a comment fix you'll have to make anyway): I'd move this to the top of the function, even before taking the lock. It's probably safe as it is currently, but I can imagine a callback that misbehaves when the server is "closed".
<!-- Reviewable comment -K2IPmCDPFzBM2gV5Z5i:15716165 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

spencerkimball committed:
Optimize gossip network against preset maximum hops for info.

Previously, we were always allowing a set number of outgoing and incoming
connections from a gossip node. Now we optimize that number based on the
total number of nodes in the network and a constant default for the maximum
number of hops that any gossip info requires to transit from origin node
to the furthest destination node.

Added a simulation step mechanism to improve the output from the gossip
simulation. Now you can see each step much more clearly, because every
server response is gated on successive simulation cycles. It's still not
perfect, but it's a big improvement and "perfect" looks like it will
require considerably more complex synchronization.

tamird commented:
Reviewed 17 of 17 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/gossip.go, line 76 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2SfmEdO6fzjP1qRFgn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3778031ad5d3ad9d4581ed6d752d9baf4520cc55/gossip/gossip.go#L76)):</sup>
constants are untyped, so this is superfluous
<!-- Reviewable comment -K2SfmEdO6fzjP1qRFgo:1482871716 -->

---

<sup>**[gossip/gossip.go, line 81 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2SfphsSqG0tJP-OZsy)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3778031ad5d3ad9d4581ed6d752d9baf4520cc55/gossip/gossip.go#L81)):</sup>
constants are untyped, so this is superfluous
<!-- Reviewable comment -K2SfphsSqG0tJP-OZsz:1482871716 -->

---

<sup>**[gossip/infostore.go, line 159 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2Sg6_R-U8brRC5mzBn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3778031ad5d3ad9d4581ed6d752d9baf4520cc55/gossip/infostore.go#L159)):</sup>
if you're using log.Fatal to validate an invariant, prefer panic, since it gives a stack trace. there are a few more instances of that here.
<!-- Reviewable comment -K2Sg6_R-U8brRC5mzBo:-1352450739 -->

---

<sup>**[gossip/simulation/network.go, line 72 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2Sfe-pmnjZ9ATOrOLq)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3778031ad5d3ad9d4581ed6d752d9baf4520cc55/gossip/simulation/network.go#L72)):</sup>
how come just one?
<!-- Reviewable comment -K2Sfe-pmnjZ9ATOrOLr:-1531612474 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 27 of 29 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[gossip/infostore.go, line 159 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2Sg6_R-U8brRC5mzBn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3778031ad5d3ad9d4581ed6d752d9baf4520cc55/gossip/infostore.go#L159)):</sup>
I thought fatal dumped stack traces. That's the way it worked in Google C++ land anyway.
<!-- Reviewable comment -K2TjBYVBYMApTk4NCiq:748807265 -->

---

<sup>**[gossip/server.go, line 169 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2IPmCDPFzBM2gV5Z5h)** ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fdc51c55e9e18c7efec4cbf22b4673a110a2fa/gossip/server.go#L169)):</sup>
This unregister callback isn't protected by a mutex as it's taken directly from the direct call to infoStore.registerCallback. So it needs to be done with the lock held. It doesn't have any interaction with the closed boolean, though I don't mind reordering them.
<!-- Reviewable comment -K2TjWt2qUpqxT1Fm5Bl:1904892876 -->

---

<sup>**[gossip/simulation/network.go, line 72 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2Sfe-pmnjZ9ATOrOLq)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3778031ad5d3ad9d4581ed6d752d9baf4520cc55/gossip/simulation/network.go#L72)):</sup>
The simulation output is simpler to decipher this way. But ultimately it doesn't matter. At some point we might make it a parameter how many there are, but I don't see much point in it.
<!-- Reviewable comment -K2TjwzeiIuLy0MiBZ26:1394762533 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[gossip/simulation/network.go, line 72 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967#-K2Sfe-pmnjZ9ATOrOLq)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3778031ad5d3ad9d4581ed6d752d9baf4520cc55/gossip/simulation/network.go#L72)):</sup>
I guess I expected there to be value in using at least two given the issues that surfaced in the acceptance test in an earlier version of this PR.
<!-- Reviewable comment -K2TmtA-qyQ2nfmTzFNc:-1885084296 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2967)
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM
